support MCP session management

zane-neo opened the pull request:
### Description
This PR is to support MCP session management to fix the issue in cluster case, the session could lose if client sends message to load balancer or random coordinators.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpMessageAction.java:
```diff
@@ -0,0 +1,109 @@
+
+                        @Override
+                        public void handleException(TransportException e) {
+                            System.out.println("got exception:" + e.getMessage());
```
remove debugging code ?

zane-neo replied:
Done

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpMessageAction.java:
```diff
@@ -0,0 +1,109 @@
+                        @Override
+                        public void handleException(TransportException e) {
+                            System.out.println("got exception:" + e.getMessage());
+                            log.error("got exception: ", e);
```
`got exception` is too simple, Make error message more readable

zane-neo replied:
Changed.

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMcpConnectionMessageStreamingAction.java:
```diff
@@ -122,31 +147,136 @@ protected void prepareRequestInternal(final String path, final String sessionId,
+                                channel.sendResponse(new BytesRestResponse(channel, e));
                             } catch (IOException ex) {
-                                log.error("Failed to send exception response to client during message handling due to IOException");
+                                log.error("Failed to get the session management index result with sessionId: {}", sessionId);
```
Error message is misleading. Here the exception should be some failure when send exception response to client. I think the original exception make sense. 

This error message should be added after link 235.

zane-neo replied:
This is the error message when getIndex request failed, so we need to change the message here

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpMessageAction.java:
```diff
@@ -0,0 +1,109 @@
+
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            Map<String, DiscoveryNode> nodes = new HashMap<>(clusterService.state().nodes().getNodes());
+            nodes.remove(clusterService.localNode().getId());
```
Why remove local node ? I see this transport action will be used in 
```
if (clusterService.localNode().getId().equals(nodeId)) {
     client
            .execute(
                MLMcpMessageAction.INSTANCE,
                new MLMcpMessageRequest(sessionId, requestBody),
                actionListener
            );
```
Can you add some comments ?

zane-neo replied:
These are test code, removing

ylwu-amzn commented:
Have you addressed all comments from last PR https://github.com/opensearch-project/ml-commons/pull/3781 ? 

Suggest add your comment on last PR if the comment already addressed in this PR

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMcpConnectionMessageStreamingAction.java:
```diff
@@ -122,31 +147,136 @@ protected void prepareRequestInternal(final String path, final String sessionId,
+                                        }
+                                    );
+                                    if (clusterService.localNode().getId().equals(nodeId)) {
+                                        client
```
Do we need this ? I see same code in else block

ylwu-amzn replied:
If you check the first commit, you can see such same code 
```
                                        client
                                            .execute(
                                                MLMcpMessageAction.INSTANCE,
                                                new MLMcpMessageRequest(sessionId, requestBody),
                                                actionListener
                                            );
```

ylwu-amzn replied:
But I see you removed such code in the second commit for `(clusterService.localNode().getId().equals(nodeId))`. I think such code not needed for local node case, right ?

zane-neo replied:
At first I thought you mean the `clusterService.localNode().getId().equals(nodeId)` in else. I removed the part you mentioned, it's not needed.

zane-neo commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpMessageDispatchedAction.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    protected void doExecute(Task task, ActionRequest request, ActionListener<AcknowledgedResponse> listener) {
+        MLMcpMessageRequest mlMcpMessageRequest = MLMcpMessageRequest.fromActionRequest(request);
+        final StreamingRestChannel channel = McpAsyncServerHolder.CHANNELS.get(mlMcpMessageRequest.getSessionId());
```
rest channels here represents persistent connections, it's ok to use them anywhere, but we need somewhere to close them, will add this logic.

zane-neo commented:
> Have you addressed all comments from last PR #3781 ?
> 
> [#3781 (comment)](https://github.com/opensearch-project/ml-commons/pull/3781#issuecomment-2837188288), I don't see setting to enable/disable this experimental feature
> 
> Suggest add your comment on last PR if the comment already addressed in this PR

Those comments will be addressed in next PR.

zane-neo commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/OpenSearchMcpServerTransportProvider.java:
```diff
@@ -96,24 +106,33 @@ public Mono<Void> closeGracefully() {
-            String result = String.format("/sse/message?sessionId=%s", sessionId);
-            return Mono.just(createHttpChunk(ENDPOINT_EVENT_TYPE, result));
+            ActionListener<IndexResponse> actionListener = ActionListener.wrap(r -> {
+                if (r != null && r.status() == RestStatus.CREATED) {
```
good catch, added else logic.

zane-neo commented on common/src/main/resources/index-mappings/ml_mcp_session_management.json:
```diff
@@ -0,0 +1,21 @@
+    "node_id": {
+      "type": "keyword"
+    },
+    "status": {
```
Values are: active and inactive. Now code has only active, created is the rest response status.

zane-neo commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpMessageAction.java:
```diff
@@ -0,0 +1,109 @@
+
+                        @Override
+                        public String executor() {
+                            return SAME;
```
Which thread will be use to handle the transport response.

zane-neo commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpMessageDispatchedAction.java:
```diff
@@ -0,0 +1,84 @@
+                    channel.sendResponse(new BytesRestResponse(channel, new Exception(e)));
+                    listener.onFailure(new Exception(e));
+                } catch (IOException ex) {
+                    log.error("Failed to send exception response to client during message handling due to IOException");
```
it's logged.

zane-neo replied:
O, I mean I'll add this log, and I've logged in the comment addressing commit.

zane-neo commented on plugin/src/main/java/org/opensearch/ml/action/mcpserver/McpAsyncServerHolder.java:
```diff
@@ -21,6 +26,8 @@ public class McpAsyncServerHolder {
         new ObjectMapper()
     );
 
+    public static Map<String, StreamingRestChannel> CHANNELS = new ConcurrentHashMap<>();
```
Based on the lifecycle of MCP protocol, we should close the channel when client disconnected, but I'm not sure if we have an approach to listen to this event. This doesn't have big impact for now so we can deprioritize this.

ylwu-amzn reviewed (dismissed):
Approved to unblock RC2. Please address all comments in new PR

ylwu-amzn commented:
CI failed
```
Suite: Test class org.opensearch.ml.rest.mcpserver.RestMLRegisterMcpToolsActionTests
  2> REPRODUCE WITH: ./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.mcpserver.RestMLRegisterMcpToolsActionTests.test_routes" -Dtests.seed=9611987A44BAA22D -Dtests.security.manager=false -Dtests.locale=fr-DZ -Dtests.timezone=SystemV/EST5EDT -Druntime.java=23
  2> java.lang.AssertionError: expected:<[POST /_plugins/_ml/mcp_tools/_register]> but was:<[POST /_plugins/_ml/mcp/tools/_register]>
        at __randomizedtesting.SeedInfo.seed([9611987A44BAA22D:DA045D84C754C8A3]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:120)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at org.opensearch.ml.rest.mcpserver.RestMLRegisterMcpToolsActionTests.test_routes(RestMLRegisterMcpToolsActionTests.java:172)
```

zane-neo commented on plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMcpConnectionMessageStreamingAction.java:
```diff
@@ -122,31 +147,136 @@ protected void prepareRequestInternal(final String path, final String sessionId,
+                                String nodeId = String.valueOf(r.getSourceAsMap().get("node_id"));
+                                DiscoveryNode node = clusterService.state().getNodes().getNodes().get(nodeId);
+                                if (node == null) {
+                                    log.error("The node:{} is no longer in the current cluster, can not handle the mcp request", r.getId());
```
Changed

zane-neo commented on plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMcpConnectionMessageStreamingAction.java:
```diff
@@ -122,31 +147,136 @@ protected void prepareRequestInternal(final String path, final String sessionId,
+                                            log
+                                                .error(
+                                                    "MCP request has been dispatched to corresponding node but peer node failed to handle it",
+                                                    e
```
done

zane-neo commented on plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMcpConnectionMessageStreamingAction.java:
```diff
@@ -122,31 +147,136 @@ protected void prepareRequestInternal(final String path, final String sessionId,
+                                    }
+                                }
+                            } else {
+                                log.error("SessionId not found in cluster");
```
done

zane-neo commented on plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMcpConnectionMessageStreamingAction.java:
```diff
@@ -122,31 +147,136 @@ protected void prepareRequestInternal(final String path, final String sessionId,
+                                            )
+                                        );
+                                } catch (IOException ex) {
+                                    log.error("Failed to send exception response to client when session ID not found in cluster state");
```
done

zane-neo commented:
> Let's try to take care of these comment since we cannot modify them later (will keep adding more as I see them):
> 
> * [initial commit for MCP server in OpenSearch #3781 (comment)](https://github.com/opensearch-project/ml-commons/pull/3781#issuecomment-2837234523)
> * https://github.com/opensearch-project/ml-commons/pull/3781/files#r2064552594
> * https://github.com/opensearch-project/ml-commons/pull/3781/files#r2064559936
> * https://github.com/opensearch-project/ml-commons/pull/3781/files#r2064666837
> * https://github.com/opensearch-project/ml-commons/pull/3781/files#r2064870075
> 
> Additionally, for the comments that we don't address, can you create issues to track them?

These comments are been address in this PR.

zane-neo commented on common/src/main/java/org/opensearch/ml/common/transport/mcpserver/requests/register/McpTool.java:
```diff
@@ -92,26 +93,26 @@ public static McpTool parse(XContentParser parser) throws IOException {
                     break;
```
Which fields? Not following.

ylwu-amzn reviewed (approved):
approve to unblock 3.0 RC2
